File: LeetCode/python3/967.py
class Solution:
    def numsSameConsecDiff(self, N, K):
        """
        :type N: int
        :type K: int
        :rtype: List[int]
        """
        ret = []
        if N == 1:
            ret.append(0)
        for i in range(1, 10):
            self.helper(i, N - 1, K, ret, i)
        return ret

    def helper(self, last, N, K, ret, val):
        if N <= 0:
            ret.append(val)
            return
        if last + K < 10:
            self.helper(last + K, N - 1, K, ret, val * 10 + last + K)
        if last - K >= 0 and K != 0:
            self.helper(last - K, N - 1, K, ret, val * 10 + last - K)

File: LeetCode/cpp/967.cpp
class Solution {
public:
    vector<int> ret;

    vector<int> numsSameConsecDiff(int N, int K) {
        if (N == 0)
            return {0};
        ret.clear();
        if (N == 1)
            ret.push_back(0);

        for (int i = 1; i <= 9; ++i)
            helper(i, N - 1, K, i);

        return ret;
    }

    void helper(int last, int N, int K, int x) {
        if (N <= 0) {
            ret.push_back(x);
            return;
        }
        if (last + K < 10)
            helper(last + K, N - 1, K, x * 10 + last + K);
        if (last - K >= 0 && K != 0)
            helper(last - K, N - 1, K, x * 10 + last - K);
    }
};